src/hotspot/share/prims/unsafe.cpp

```diff
@@ -783,30 +783,20 @@ UNSAFE_ENTRY(void, Unsafe_Park(JNIEnv *env, jobject unsafe, jboolean isAbsolute,
 
 UNSAFE_ENTRY(void, Unsafe_Unpark(JNIEnv *env, jobject unsafe, jobject jthread)) {
   if (jthread != nullptr) {
-    // ThreadsListHandle.cv_internal_thread_to_JavaThread() would normally
-    // be used for this conversion from jthread -> JavaThread*, but we've
-    // seen performance issues with very large numbers of threads due to
-    // the ThreadsListHandle.includes() call in the conversion function.
-    //
-    // So we verify that java_lang_Thread::thread(thread_oop) != nullptr
-    // before and after creating tlh below and that lets us know that thr
-    // is protected by tlh without a tlh.includes() call.
-    //
     oop thread_oop = JNIHandles::resolve_non_null(jthread);
-    if (java_lang_Thread::thread(thread_oop) != nullptr) {
-      // Try to capture the live JavaThread in a ThreadsListHandle:
-      ThreadsListHandle tlh;
-      JavaThread* thr = nullptr;
-      if ((thr = java_lang_Thread::thread(thread_oop)) != nullptr) {
-        // The still live JavaThread is protected by the ThreadsListHandle
-        // so it is safe to access.
-        assert(tlh.includes(thr), "must be");
-        Parker* p = thr->parker();
-        HOTSPOT_THREAD_UNPARK((uintptr_t) p);
-        p->unpark();
-      }
-    } // ThreadsListHandle is destroyed here.
-  }
+    // Get the JavaThread* stored in the java.lang.Thread object _before_
+    // the embedded ThreadsListHandle is constructed so we know if the
+    // early life stage of the JavaThread* is protected.
+    FastThreadsListHandle ftlh(thread_oop, java_lang_Thread::thread(thread_oop));
+    JavaThread* thr = ftlh.protected_java_thread();
```

Comment:
I find this obscures things more than it aids understanding in this case. You don't need a special FTLH to get the desired semantics here.

Reply:
Note this is just a comment I'm not requesting you change it back.

Reply:
Thanks for making it clear that the above is just a comment.

I rather like the fact that this is best encapsulation of this crazy performance
logic and it allows us to put (most of) the logic over in the threadSMR source
files where it belongs.

Of all the possible solutions we have tried out, I like this one the most and I
think that it's the least hacky. I wouldn't quite say that this solution "sings",
but it comes the closest (from my POV)...